src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs

```diff
@@ -397,18 +398,18 @@ matchNonVarToPattern tools first second
             Equals.makeEvaluateTermsToPredicateSubstitution tools first second
         let
             -- We're only interested in substitutions involving first's
-            -- variables
-            -- here, and there are no free variables in the RHS, so we're
-            -- coverting everything else to predicates.
+            -- variables here, and we're converting everything else to
+            -- predicates.
             -- TODO: Make a function for this.
-            (finalPredicate, _proof) =
-                makeAndPredicate
-                    predicate
-                    (substitutionToPredicate substitution)
+            leftVars = pureFreeVariables Proxy first
+            (leftSubst, rightSubst) =
+                span ((`elem` leftVars).fst) substitution
+            finalPredicate = PredicateSubstitution.toPredicate
+                PredicateSubstitution { predicate, substitution = rightSubst }
         return . return $ -- MonadCounter m => m (Maybe a)
             PredicateSubstitution
                 { predicate = finalPredicate
-                , substitution = []
+                , substitution = leftSubst
```

Comment:
Should add a test or two for this, especially in light of the `span` error above.